File: android/src/main/java/expo/modules/stickyview/ExpoStickyView.kt
package expo.modules.stickyview

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.view.ViewParent
import android.widget.ScrollView
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView
import com.facebook.react.views.scroll.ReactScrollView
import com.facebook.react.views.scroll.ReactScrollViewHelper
import com.facebook.react.views.scroll.ScrollEventType
import expo.modules.kotlin.viewevent.EventDispatcher
import kotlin.math.abs
import kotlin.math.round

fun View.getParentView(): View? {
  val parent = this.parent
  return if (parent is View) parent else null
}

fun View.findClosestScrollableAncestor(): ReactScrollView? {
  var ancestor: ViewParent? = this.parent
  while (ancestor is ViewGroup) {
    if (ancestor is ReactScrollView) {
      return ancestor
    }
    ancestor = ancestor.parent
  }
  return null
}

fun View.getLocationInWindow(): IntArray {
  val location = IntArray(2)
  this.getLocationInWindow(location)
  return location
}

fun View.getLocationRelativeToViewWithoutTranslation(view: View): IntArray {
  val location = IntArray(2)
  this.getLocationInWindow(location)
  location[0] -= translationX.toInt()
  location[1] -= translationY.toInt()
  val viewLocation = view.getLocationInWindow()
  location[0] -= viewLocation[0]
  location[1] -= viewLocation[1]
  return location
}

fun Context.dpToPx(dp: Double): Double {
  return (dp * this.resources.displayMetrics.density)
}

fun Context.pxToDp(px: Double): Double {
  return (px / this.resources.displayMetrics.density)
}

sealed class StickyMode {
  data class Top(val offset: Double) : StickyMode()
  data class Bottom(val offset: Double) : StickyMode()
}

class ExpoStickyView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  val onStickyChange by EventDispatcher()
  private var scrollableAncestor: ReactScrollView? = null
  private var stickyMode: StickyMode = StickyMode.Top(0.0)
  private val scrollListener: ReactScrollViewHelper.ScrollListener = object : ReactScrollViewHelper.ScrollListener {
    override fun onScroll(
      scrollView: ViewGroup?,
      scrollEventType: ScrollEventType?,
      xVelocity: Float,
      yVelocity: Float
    ) {
      if (scrollView === scrollableAncestor) {
        applySticky()
      }
    }
    override fun onLayout(scrollView: ViewGroup?) {
      applySticky()
    }

  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    this.scrollableAncestor = findClosestScrollableAncestor()
    ReactScrollViewHelper.addScrollListener(this.scrollListener)
  }

  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    ReactScrollViewHelper.removeScrollListener(this.scrollListener)
    this.scrollableAncestor = null
  }

  private fun applySticky() {
    // Calculate the position of this view relative to the scrollable ancestor
    val parentView = getParentView()
    val scrollableAncestor = this.scrollableAncestor
    if (parentView == null || scrollableAncestor == null) {
      return
    }
    when(val stickyMode = this.stickyMode) {
      is StickyMode.Top -> applyStickyTop(scrollableAncestor, parentView, stickyMode.offset)
      is StickyMode.Bottom -> applyStickyBottom(scrollableAncestor, parentView, stickyMode.offset)
    }
  }
  private fun applyStickyTop(scrollableAncestor: ScrollView, parentView: View, offset: Double) {
    val relativePosition = getLocationRelativeToViewWithoutTranslation(scrollableAncestor)
    val relativeY = relativePosition[1]
    val maxTranslation = parentView.height - this.height - this.top
    val shouldStick = relativeY < offset
    if (shouldStick) {
      val pendingTranslation = offset - relativeY
      this.translationY = pendingTranslation.toFloat().coerceAtMost(maxTranslation.toFloat())
    } else {
      this.translationY = 0.0f
    }
    onStickyChange(mapOf(
        "isStuck" to shouldStick,
        "currentFloatDistance" to round(context.pxToDp(abs(this.translationY.toDouble()))),
        "maxFloatDistance" to round(context.pxToDp(abs(maxTranslation.toDouble())))
    ))
  }

  private fun applyStickyBottom(scrollableAncestor: ScrollView, parentView: View, offset: Double) {
    val relativePosition = getLocationRelativeToViewWithoutTranslation(scrollableAncestor)
    val relativeY = relativePosition[1]
    val scrollableAncestorHeight = scrollableAncestor.height

    val relativeBottom = relativeY + this.height
    // The max offset we can have is this.top to float to the top of the parent
    val minTranslation = -this.top.toDouble()

    val shouldStick = relativeBottom > scrollableAncestorHeight - offset
    if (shouldStick) {
      val delta = scrollableAncestorHeight - offset - relativeBottom
      this.translationY = delta.coerceAtLeast(minTranslation).toFloat()
    } else {
      this.translationY = 0.0f
    }
    onStickyChange(mapOf(
      "isStuck" to shouldStick,
      "currentFloatDistance" to round(context.pxToDp(abs(this.translationY.toDouble()))),
      "maxFloatDistance" to round(context.pxToDp(abs(minTranslation))),
    ))
  }

  fun setStickyTop(offset: Double) {
    this.stickyMode = StickyMode.Top(context.dpToPx(offset))
    applySticky()
  }
  fun setStickyBottom(offset: Double) {
    this.stickyMode = StickyMode.Bottom(context.dpToPx(offset))
    applySticky()
  }
}

File: ios/ExpoStickyView.swift
import ExpoModulesCore
import UIKit

enum StickyMode {
  case top(offset: CGFloat)
  case bottom(offset: CGFloat)
}

class ExpoStickyView: ExpoView {
  let onStickyChange = EventDispatcher()
  private var stickyMode: StickyMode = .top(offset: 0)
  private weak var scrollableAncestor: UIScrollView?
  private var scrollObserver: NSKeyValueObservation?
  private var layoutObserver: NSKeyValueObservation?
  private var untransformedFrame: CGRect = CGRectZero()
    
  override func didMoveToWindow() {
    super.didMoveToWindow()
    if window != nil {
      attachToScrollableAncestor()
    } else {
      detachFromScrollableAncestor()
    }
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    self.untransformedFrame = CGRectMake(self.frame.origin.x, self.frame.origin.y - self.transform.ty, self.frame.width, self.frame.height)
    self.applySticky()
  }

  private func attachToScrollableAncestor() {
    guard let ancestor = findClosestScrollableAncestor() else { return }
    scrollableAncestor = ancestor

    scrollObserver = ancestor.observe(\.contentOffset, options: [.new]) { [weak self] _, _ in
      self?.applySticky()
    }
    layoutObserver = ancestor.observe(\.frame, options: [.new]) { [weak self] _, _ in
      self?.applySticky()
    }
    applySticky()
  }

  private func detachFromScrollableAncestor() {
    scrollObserver?.invalidate()
    layoutObserver?.invalidate()
    scrollObserver = nil
    layoutObserver = nil
    scrollableAncestor = nil
  }

  private func findClosestScrollableAncestor() -> UIScrollView? {
    var ancestor: UIView? = self.superview
    while let view = ancestor {
      if let scroll = view as? UIScrollView {
        return scroll
      }
      ancestor = view.superview
    }
    return nil
  }

  private func getLocationRelativeToView(_ view: UIView) -> CGPoint {
    let myFrame = self.convert(self.bounds, to: view)
    return CGPoint(x: myFrame.origin.x - self.transform.tx, y: myFrame.origin.y - self.transform.ty)
  }

  private func applySticky() {
    guard let ancestor = scrollableAncestor, let parent = self.superview else { return }
    switch stickyMode {
    case .top(let offset):
      applyStickyTop(scrollableAncestor: ancestor, parentView: parent, offset: offset)
    case .bottom(let offset):
      applyStickyBottom(scrollableAncestor: ancestor, parentView: parent, offset: offset)
    }
  }

  private func applyStickyTop(scrollableAncestor: UIScrollView, parentView: UIView, offset: CGFloat) {
    let relativePosition = getLocationRelativeToView(scrollableAncestor)
    let relativeY = relativePosition.y - scrollableAncestor.contentOffset.y
    let maxTranslation = CGFloat(parentView.bounds.height - self.bounds.height - self.untransformedFrame.origin.y)
    let shouldStick = relativeY < offset
      
//    NSLog("Bounds \(self.bounds.origin.y) RelativeY: \(relativePosition.y), ContentOffset: \(scrollableAncestor.contentOffset.y) OriginalY: \(self.untransformedFrame.origin.y)")
    if shouldStick {
      let pendingTranslation = offset - relativeY
      self.transform = CGAffineTransform(translationX: 0, y: min(pendingTranslation, maxTranslation))
    } else {
      self.transform = .identity
    }
    onStickyChange([
      "isStuck": shouldStick,
      "currentFloatDistance": abs(self.transform.ty),
      "maxFloatDistance": abs(maxTranslation)
    ])
  }

  private func applyStickyBottom(scrollableAncestor: UIScrollView, parentView: UIView, offset: CGFloat) {
    let relativePosition = getLocationRelativeToView(scrollableAncestor)
    let relativeY = relativePosition.y - scrollableAncestor.contentOffset.y
    let scrollableAncestorHeight = scrollableAncestor.bounds.height
    let relativeBottom = relativeY + self.bounds.height
    let minTranslation = -self.untransformedFrame.origin.y
    let shouldStick = relativeBottom > scrollableAncestorHeight - offset
    if shouldStick {
      let delta = scrollableAncestorHeight - offset - relativeBottom
      self.transform = CGAffineTransform(translationX: 0, y: max(delta, minTranslation))
    } else {
      self.transform = .identity
    }
    onStickyChange([
      "isStuck": shouldStick,
      "currentFloatDistance": abs(self.transform.ty),
      "maxFloatDistance": abs(minTranslation)
    ])
  }

  // MARK: - Props
  func setStickyTop(_ offset: Double) {
    stickyMode = .top(offset: CGFloat(offset))
    applySticky()
  }

  func setStickyBottom(_ offset: Double) {
    stickyMode = .bottom(offset: CGFloat(offset))
    applySticky()
  }
}
